### DISCUSSION TITLE
Object serialization / saving an object to filesystem?

### DISCUSSION THREAD

**Discussion by IcedRooibos** [micropython/micropython#10790] [General]
Is there any recommended/general purpose way to save objects to the filesystem at the moment? Or should I go ahead and write custom code based on how my particular objects are structured?

Here's a very basic example to show what I'd like to do

```

class refObj():
    def __init__(self, v1, v2, v3, v4, rObj=None):
        self.var1 = v1
        self.var2 = v2
        self.var3 = v3
        self.var4 = v4
        self.rObj = rObj
    def doStuff(self):
        print(self.var1+self.var2+self.var3+self.var4)
        if self.rObj != None:
            self.rObj.doStuff()

        
a = refObj(1,2,3,4)
b = refObj(5,6,7,8,a)
c = refObj(9,10,11,12,b)

# Save above objects to filesystem to use later (in practice they
# will be defined through user input)

# write object to a file
# How?

# Stuff happens, power loss etc
del a
del b
del c

# Load above objects on startup
# How?

a.doStuff()
b.doStuff()
c.doStuff()

```

**Comment by cve2022** [comment]
What about [json](https://docs.micropython.org/en/latest/library/json.html)?

**Comment by brave-ulysses** [comment]
https://pypi.org/project/micropython-pickle/

**Comment by IcedRooibos** [comment]
@cve2022 

Thank you! I tried to use `json.dumps` on `object.__dict__` but it doesn't handle recursion natively i.e. another object is part of `object.__dict__`. This means I would have to write custom code to handle that. I just asked this question to make sure that I wasn't missing some library that was already written and tested.

@brave-ulysses 

Thank you, I did see that earlier but unfortunately it looks like it's not developed yet.

From the link:
`This is a dummy implementation of a module for MicroPython standard library. It contains zero or very little functionality`

**Reply by bixb922** [reply]
A updated version of pickle is in  micropython-lib (https://github.com/micropython/micropython-lib/tree/master/python-stdlib/pickle).

**Reply by IcedRooibos** [reply]
That's awesome, thank you @bixb922!

**Comment by peterhinch** [comment]
[This doc](https://github.com/peterhinch/micropython-samples/blob/master/SERIALISATION.md) lists several options for serialisation with the benefits and drawbacks of each.

**Reply by IcedRooibos** [reply]
Thank you!
